### ISSUE TITLE
Add @RootContext in @ViewGroup

### ISSUE THREAD

**Issue by LouisCo** [excilys/androidannotations#1540]
Could `@RootContext` be supported in `@ViewGroup` to make us able to use the type safe real context? In my case, I would like to access my Service instance from my `@ViewGroup`, showed in `WindowManager`

**Comment by WonderCsabo** [comment]
I am not sure. The reason for the `@RootContext` existence is there is no way to get the `Context` in your beans, because they are only added in the generated class. In `View`s, you can just call `getContext()` and cast.

**Comment by LouisCo** [comment]
But isn't AA dedicated to avoid all theses casts as it does with `@ViewById`?

**Comment by WonderCsabo** [comment]
Well, yeah, it does. I am just saying that `@RootContext` was not designed for this purpose originally. I am not sure we should extend it or not. @yDelouis @dodgex ?

**Comment by dodgex** [comment]
I'm not sure. Is there a reason to not extend it? As long as `getContext()` returns the `Service` or `Activity` (i'm not sure here as i never required more than a plain `Context` in my view), that case this would be a valid request. and it is one of AAs benefits to not have to take care of stuff like that.

**Comment by WonderCsabo** [comment]
Actually i am not sure that a `View` should be aware of the exact `Activity` or `Service`. Sounds like a bad design.

**Comment by dodgex** [comment]
In terms of reusability this is true. But in some cases/projects using a bad design like this is totaly valid. e.g. In the view is only created to separate some Code to not have to many LoC in one File. 

While He cool create an Interface for callbacks to the Service this might again mean some additional boilerplate code. 

I see a point on both sides. Not sure which one has the better point.

**Comment by WonderCsabo** [comment]
Another vote is needed. :) @yDelouis WDYT?

**Comment by LouisCo** [comment]
In terms of reusability, but if my ViewGroup is only intended to be used with WindowManager and a specific Service? Maybe being able to inject objects (for example an object implenting a contract interface) instead would be better? For now, this second option is hardly possible in AA, it requires additional boilerplate.

**Comment by yDelouis** [comment]
I think it could be good if we also allow to inject interfaces with `RootContext`. Here is an example : 
````
java
@ViewGroup
public class MyViewGroup extends ViewGroup {

     public interface Callback {
           void onEvent()
     }

    @RootContext Callback callback;
}
`````
This way, we allow to inject a `Context` without knowing exactly which context it is. And It enables @LouisCAD's use case.
The `Context` will be injected only if implements the interface.

**Comment by WonderCsabo** [comment]
I am really not sure to allow implementing any kind of interface. From an annotation named `@RootContext`, i would expect to inject a `Context` object. I am sorry, but i feel this kind of usage abuses the original goal of the annotation. A `Context` is not a `Callback`. Yeah, actually a `Context` subclass may implement it, but then we are not using it as a `Context`.

**Comment by LouisCAD** [comment]
@yDelouis Your idea I great, I often need to get dependencies in my Beans, and it makes me do bad looking code. This way, it could help, but as @WonderCsabo , I think `@RootContext` would be more for Contexts. Something like `@Dependency` would be better in my opinion.

**Comment by WonderCsabo** [comment]
@LouisCAD `@Bean` should be used for that. Also for @yDelouis' case, i would use `EventBus` or `Otto`.

BTW, i am now OK to extend `@RootContext` support for `@EView`, `@EViewGroup` and even `@EFragment`, with the current restriction for the annotated element type. It should not hurt anybody and easy to implement.
However, i have to add this is not typesafe at all, as opposed you wrote in the first post. We just do the same cast what you are doing currently in the code.

**Comment by WonderCsabo** [comment]
@yDelouis should we enable it or not? The `@RootContext` javadoc says it is clearly for `@EBean`s. I think users would find pretty confusing if we re-purpose an annotation.

**Comment by LouisCAD** [comment]
@WonderCsabo But `@EBean` can't inject non annotated dependencies, neither already existing dependencies...

**Comment by WonderCsabo** [comment]
@LouisCAD unfortunately i cannot understand how does that relate to my comment.

**Comment by LouisCAD** [comment]
@WonderCsabo My comment related to:
> `@Bean` should be used for that.
Sometimes, I need to inject already existing dependencies, or non annotated dependencies, but I can't add constructor parameters with AA. that's why I would like to be able to inject dependency provider, if possible

**Comment by WonderCsabo** [comment]
OK, please add example code how `@RootContext` in `@EView` would help you.

**Comment by dodgex** [comment]
> Sometimes, I need to inject already existing dependencies, or non annotated dependencies, but I can't add constructor parameters with AA. that's why I would like to be able to inject dependency provider, if possible

sounds like: https://github.com/excilys/androidannotations/issues/1520

**Comment by LouisCAD** [comment]
@dodgex It's like #1520 ! @yDelouis comment made me think of that.
@WonderCsabo I have a use case for `@RootContext` in `@EViewGroup`, I'll show you a code example tonight.

**Comment by WonderCsabo** [comment]
@yDelouis should we add this?

**Comment by yDelouis** [comment]
Yes, we can add this. I don't think it is necessary for 4.0 though.

**Comment by WonderCsabo** [comment]
I agree. Contributions are welcome, adding this is pretty straightforward.
